tools/outline_proxy_controller/outline_proxy_controller.cpp

```diff
@@ -0,0 +1,571 @@
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+#include "logger.h"
+#include "outline_proxy_controller.h"
+
+using namespace std;
+using namespace outline;
+
+extern Logger logger;
+
+string OutlineProxyController::getParamValueInResult(const string resultString,
+                                                     const string param) {
+  auto paramPosition = resultString.find(param);
+  if (paramPosition == string::npos) throw std::invalid_argument("param not found");
+
+  auto valuePosition = paramPosition + param.length() + resultDelimiter.length();
+  if (valuePosition > resultString.length()) return "";
+
+  auto delimiterPosition = resultString.find(resultDelimiter, valuePosition);
+  delimiterPosition =
+      (delimiterPosition != string::npos) ? delimiterPosition : resultString.length();
+
+  return resultString.substr(valuePosition, delimiterPosition - valuePosition);
+}
+
+OutputAndStatus OutlineProxyController::executeIPRoute(const SubCommand args) {
+  return executeCommand(IPRouteCommand, args);
+}
+
+OutputAndStatus OutlineProxyController::executeIPLink(const SubCommand args) {
+  return executeCommand(IPLinkCommand, args);
+}
+
+OutputAndStatus OutlineProxyController::executeIPTunTap(const SubCommand args) {
+  return executeCommand(IPTunTapCommand, args);
+}
+
+OutputAndStatus OutlineProxyController::executeCommand(const std::string commandName,
+                                                       const SubCommand received_args) {
+  array<char, 128> buffer;
+  string result;
+  string cmd = commandName;
+
+  SubCommand args = received_args;
+
+  while (!args.empty()) {
+    auto curArg = args.front();
+    cmd += " " + (curArg.first) + " " + (curArg.second);
+    args.pop();
+  }
+
+  cmd += c_redirect_stderr_into_stdout;
+
+  FILE* pipe = popen(cmd.c_str(), "r");
+  if (!pipe) throw std::runtime_error("failed to run " + commandName + " command!");
+
+  while (!feof(pipe)) {
+    if (fgets(buffer.data(), 128, pipe) != nullptr) result += buffer.data();
+  }
+
+  return make_pair(result, WEXITSTATUS(pclose(pipe)));
+}
+
+OutlineProxyController::OutlineProxyController() {
+  detectBestInterfaceIndex();
+  addOutlineTunDev();
+  setTunDeviceIP();
+}
+
+void OutlineProxyController::addOutlineTunDev() {
+  if (!outlineTunDeviceExsits()) {
+    // first we check if it exists and not try to add it
+    SubCommand createTunDeviceCommand;
```

Comment:
ip has different sub commands, like route, link, add, del,  which then each takes different args. The subcommand always should come first. I believe arranging them this way is less confusing, rather than treating them as bunch of flat args.